builtins: add levenshtein_less_equal function

Azanul opened the pull request:
This adds the levenshtein_less_equal function with two overloads:
one with default costs and one with custom costs.

informs: #56820
Release note (sql change): Implemented levenshtein_less_equal(string, string, int)
and levenshtein_less_equal(string, string, int, int, int, int) built-in
functions.

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=104649) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=104649) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=104649) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/104649)
<!-- Reviewable:end -->

mgartner reviewed (changes requested):
Thanks for the contribution!

We'll want extensive unit tests for the two new library functions. Take a look at the existing tests in `pkg/util/fuzzystrmatch/leven_test.go` for some inspiration on how to write these.

Please add some tests for this special case mentioned in the [Postgres docs](https://www.postgresql.org/docs/current/fuzzystrmatch.html#id-1.11.7.26.7):

> If max_d is negative then the behavior is the same as levenshtein.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-:-NXVqp3kCAQ-Erxqbvti:b-6j5dsi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azanul, @jordanlewis, @otan, and @rafiss)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NXVrEVVE3nnVjmDbCI9):*
nit: fix `levenstein` misspelling and your commit message should be formatted like this:

```
builtins: add levenshtein_less_than

<explanation of the change>

Release note (sql change): <explanation of the new builtins>
```

We have some guidelines on commit messages here: https://wiki.crdb.io/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages
___
*[`-- commits` line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NXVrnMLAx_ipLLPWarj):*
The tests should be in the same commit as the implementation.
___
*[`pkg/sql/sem/builtins/builtins.go` line 3878 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NXVregv2kYslnuxg2rn:-NXVregwD3APN3NWi282:brprk19) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfc4dad72bbc840dc63d8ae597ea96dbdb117661/pkg/sql/sem/builtins/builtins.go#L3878)):*
> ```Go
> 				s, t := string(tree.MustBeDString(args[0])), string(tree.MustBeDString(args[1]))
> 				d := int(tree.MustBeDInt(args[2]))
> 				const maxLen = 255 // Same limit as Postgres. Ref: https://github.com/postgres/postgres/blob/53ea2b7ad050ce4ad95c89bb55197209b65886a1/src/backend/utils/adt/levenshtein.c#L26
> ```

nit: maybe put the comment on the line above to make it a bit less wide?
___
*[`pkg/util/fuzzystrmatch/leven.go` line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NXVqZAKCGtlvtIp9Ur8:-NXVqZAKCGtlvtIp9Ur9:b-r7zoj9) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfc4dad72bbc840dc63d8ae597ea96dbdb117661/pkg/util/fuzzystrmatch/leven.go#L78)):*
> ```Go
> 
> // LevenshteinDistance is the standard Levenshtein distance where the cost of
> // insertion, deletion and substitution are all one.
> ```

nit: The costs are variable, so this comment looks wrong.
___
*[`pkg/util/fuzzystrmatch/leven.go` line 136 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NXVqd7s7bNA5hdSscz0:-NXVqd7s7bNA5hdSscz1:b-eo4ffr) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfc4dad72bbc840dc63d8ae597ea96dbdb117661/pkg/util/fuzzystrmatch/leven.go#L136)):*
> ```Go
> 
> 	// Previous and current rows of notional array.
> 	// Uses the PostgreSQL functions palloc and pfree instead of the
> ```

I don't understand this comment. Is it correct?
___
*[`pkg/util/fuzzystrmatch/leven.go` line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NXVqijP0awPM5bEpyAq:-NXVqijP0awPM5bEpyAr:b-3gi8c3) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfc4dad72bbc840dc63d8ae597ea96dbdb117661/pkg/util/fuzzystrmatch/leven.go#L192)):*
> ```Go
> 			}
> 
> 			// Take the one with minimum cost
> ```

nit: Here and in all other comments - can you please end them with a period, `.`?
___
*[`pkg/sql/logictest/testdata/logic_test/fuzzystrmatch` line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NXVsdXE0oPPeXBmw3sY:-NXVsdXE0oPPeXBmw3sZ:b-d4p9je) ([raw file](https://github.com/cockroachdb/cockroach/blob/585f032a689c8f1c329687351ce7e57c4aaf2c65/pkg/sql/logictest/testdata/logic_test/fuzzystrmatch#L47)):*
> ```
> NULL   a       NULL  NULL
> a      NULL    NULL  NULL
> NULL   NULL    NULL  NULL
> ```

Can you add some more test cases here too? Some ideas:

* ASCII strings where the levenshtein distance is less than the given max
* ASCII strings that are identical


<!-- Sent from Reviewable.io -->

Azanul commented:
@mgartner Thanks for the review, I've updated the code and commit message accordingly. Please do let me know if I've missed something or if any other changes are required.

mgartner reviewed (changes requested):
Getting closer! I left another round of comments.

> Reviewed 2 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-:-NYPK30W3o5iplB9tkmm:b46pzns)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azanul, @jordanlewis, @otan, and @rafiss)

___
*[`-- commits` line 4 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NYPCRZb60Ig5cAuFony):*
You can omit the `Release justifiction` line - we only require it for patches to previously released major versions.
___
*[`pkg/util/fuzzystrmatch/leven.go` line 75 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NYPJUQ83OrQrwGikos0:-NYPJUQ83OrQrwGikos1:b-i0uowd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b87fdea1f382116ea7d29025e4ba2ca46512d4c/pkg/util/fuzzystrmatch/leven.go#L75)):*
> ```Go
> // insertion, deletion and substitution are all one and we care about distances less than
> // or equal to maxDist.
> func LevenshteinLessThanDistance(source, target string, maxD int) int {
> ```

name this `LevenshteinLessEqualDistance`
___
*[`pkg/util/fuzzystrmatch/leven.go` line 81 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NYPJXTJ7z64xWTBBuUe:-NYPJXTJ7z64xWTBBuUf:b-98elea) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b87fdea1f382116ea7d29025e4ba2ca46512d4c/pkg/util/fuzzystrmatch/leven.go#L81)):*
> ```Go
> // LevenshteinLessThanDistanceWithCost is the Levenshtein distance between the source and
> // the target with the given costs of insertion, deletion and substitution.
> func LevenshteinLessThanDistanceWithCost(source, target string, insCost, delCost, subCost, maxDist int) int {
> ```

name this `LevenshteinLessEqualDistanceWithCost`.
___
*[`pkg/util/fuzzystrmatch/leven.go` line 104 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NYPDQun2bi0ep0ydBsm:-NYPDQun2bi0ep0ydBsn:b-qpud1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b87fdea1f382116ea7d29025e4ba2ca46512d4c/pkg/util/fuzzystrmatch/leven.go#L104)):*
> ```Go
> 		}
> 		if minTheoDist > maxDist {
> 			fmt.Println("xxx", minTheoDist)
> ```

Looks like you forgot to remove this `fmt.Println`
___
*[`pkg/util/fuzzystrmatch/leven.go` line 139 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NYPFGxZ5005ZOMMbnDG:-NYPFGxZ5005ZOMMbnDH:b89ro6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b87fdea1f382116ea7d29025e4ba2ca46512d4c/pkg/util/fuzzystrmatch/leven.go#L139)):*
> ```Go
> 	lenT++
> 
> 	// Initialize previous and current rows of notional array.
> ```

It'd be helpful to explain the algorithm briefly. I believe this is the Wagner-Fischer algorithm, correct? It'd be good to mention that and maybe link to a reference of it, like https://en.wikipedia.org/wiki/Wagner%E2%80%93Fischer_algorithm.
___
*[`pkg/util/fuzzystrmatch/leven.go` line 150 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NYPF1gJ0qbwi41oWwUa:-NYPF1gJ0qbwi41oWwUb:b-p8dtvx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b87fdea1f382116ea7d29025e4ba2ca46512d4c/pkg/util/fuzzystrmatch/leven.go#L150)):*
> ```Go
> 	// Loop through rows of the notional array.
> 	for i := 1; i < lenT; i++ {
> 		fmt.Println(prev)
> ```

Remove `fmt.Println`.
___
*[`pkg/util/fuzzystrmatch/leven.go` line 175 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NYPFunL-VYMvo350czU:-NYPFunL-VYMvo350czV:b-ptqp1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b87fdea1f382116ea7d29025e4ba2ca46512d4c/pkg/util/fuzzystrmatch/leven.go#L175)):*
> ```Go
> 		// fast-path to handle the (fairly common) case where no multibyte
> 		// characters are in the mix.  The fast-path is entitled to assume
> 		// that if s_char_len is not initialized then BOTH strings contain
> ```

I don't see the fast path case, nor any `s_char_len`. 
___
*[`pkg/util/fuzzystrmatch/leven.go` line 265 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NYPG8j91UEzxMjSjafN:-NYPG8j91UEzxMjSjafO:b-r9qn4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b87fdea1f382116ea7d29025e4ba2ca46512d4c/pkg/util/fuzzystrmatch/leven.go#L265)):*
> ```Go
> 	// current row, that's where we'll find it.
> 	fmt.Println(prev)
> 	fmt.Println(startColumn, stopColumn)
> ```

Remove `fmt.Println`s. 
___
*[`pkg/sql/logictest/testdata/logic_test/fuzzystrmatch` line 40 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NYPIHRR5yY4-AMzjgXh:-NYPIHRR5yY4-AMzjgXi:bh672at) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b87fdea1f382116ea7d29025e4ba2ca46512d4c/pkg/sql/logictest/testdata/logic_test/fuzzystrmatch#L40)):*
> ```
> 
> query TTII
> SELECT a, b, levenshtein_less_than(a, b, 3), levenshtein_less_than(a, b, 2, 3, 4, 12) FROM fuzzystrmatch_table ORDER BY id
> ```

`levenshtein_less_than` => `levenshtein_less_equal`
___
*[`pkg/sql/logictest/testdata/logic_test/fuzzystrmatch` line 49 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NYPHnL03t1NCFxtmjPV:-NYPHnL03t1NCFxtmjPW:b-rvc40c) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b87fdea1f382116ea7d29025e4ba2ca46512d4c/pkg/sql/logictest/testdata/logic_test/fuzzystrmatch#L49)):*
> ```
> NULL   NULL    NULL  NULL
> apple  apple   0     0
> a      abcd    3     6
> ```

You need to insert these values into the table `fuzzystrmatch_table`. That'll require updating all the other tests that use this table. You can automatically do that with the `--rewrite` test flag:

```
./dev test pkg/sql/logictest/tests/local -f TestLogic_fuzzystrmatch --rewrite
```
___
*[`pkg/util/fuzzystrmatch/leven_test.go` line 215 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NYPJP-s1QzBd6sfIcXA:-NYPJP-tCdETHnfAPVXg:bdwiosh) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b87fdea1f382116ea7d29025e4ba2ca46512d4c/pkg/util/fuzzystrmatch/leven_test.go#L215)):*
> ```Go
> }
> 
> func TestLevenshteinLessThanDistance(t *testing.T) {
> ```

name this `TestLevenshteinLessEqualDistance`
___
*[`pkg/util/fuzzystrmatch/leven_test.go` line 236 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NYPIclJAPtLJfe0EXgW:-NYPIclJAPtLJfe0EXgX:b-sziq27) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b87fdea1f382116ea7d29025e4ba2ca46512d4c/pkg/util/fuzzystrmatch/leven_test.go#L236)):*
> ```Go
> 		},
> 
> 		// Test cases for LevenshteinDistance greater than max distance.
> ```

Can you add a test in this section where the Levenshtein distacne is greater than the max distance by 2 or more, yet `MaxDistance+1` is expected. Like `levenshtein_less_equal('aaaa', 'xxxx', 2)` => 3.
___
*[`pkg/util/fuzzystrmatch/leven_test.go` line 295 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NYPJIHe2WSsZsslyaVj:-NYPJIHe2WSsZsslyaVk:btg3632) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b87fdea1f382116ea7d29025e4ba2ca46512d4c/pkg/util/fuzzystrmatch/leven_test.go#L295)):*
> ```Go
> }
> 
> func TestLevenshteinLessThanDistanceWithCost(t *testing.T) {
> ```

name this `TestLevenshteinLessEqualDistanceWithCost`


<!-- Sent from Reviewable.io -->

Azanul commented:
> Can you add a test in this section where the Levenshtein distacne is greater than the max distance by 2 or more, yet `MaxDistance+1` is expected. Like `levenshtein_less_equal('aaaa', 'xxxx', 2)` => 3.

This one already exists
```
{
	Source:      "vacaville",
	Target:      "fairfield",
	MaxDistance: 5,
	Expected:    6,
},
```

charlespnh commented:
Hey @Azanul, it looks like there haven't been much update on this and I want to help contribute to the PR. Aside from build break and merge conflict, it seems like feedbacks have been addressed... If you're unable to work on the issue at the moment, let me know how I can help with this one!

Azanul commented:
> Hey @Azanul, it looks like there haven't been much update on this and I want to help contribute to the PR. Aside from build break and merge conflict, it seems like feedbacks have been addressed... For now I can't work on the same PR since I don't have permission. If you're unable to work on the issue at the moment, let me know how I can help with this one!

Thanks a lot for bringing my attention here. I'll see if I can resolve issues here this weekend, if not I'll give you permission to the branch to do the same.

Azanul commented:
@charlespnh I don't think I can manage to resolve this issues here hence I've invited you to collaborate. Do let me know if you're unable to makes changes.

charlespnh commented:
Hi @mgartner, whenever you're available can you take a look at the PR again? @Azanul and I have addressed the feedbacks, and looks like the builds and tests pass as well (although CI doesn't seem to pick up the issue number referenced...). Thanks!

charlespnh commented:
friendly ping! @mgartner

charlespnh committed:
builtins: add levenshtein_less_equal function

This adds the levenshtein_less_equal function with two overloads:
one with default costs and one with custom costs.

Release note (sql change): Implemented levenshtein_less_equal(string, string, int)
and levenshtein_less_equal(string, string, int, int, int, int) built-in
functions.

Co-authored-by: Charles Nguyen <phucnh402@gmail.com>
Co-authored-by: Azanul <azanulhaque@gmail.com>
Co-authored-by: Rafi Shamim <rafi@cockroachlabs.com>

mgartner commented:
@charlespnh Apologies for the delay! I lost track of this. I'll take a look in the next couple of days, so expect an update soon.

Azanul commented:
> @charlespnh Apologies for the delay! I lost track of this. I'll take a look in the next couple of days, so expect an update soon.

Hi there

mgartner reviewed (dismissed):
> Reviewed 1 of 6 files at r6, 2 of 6 files at r7, 3 of 3 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104649)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @Azanul, @jordanlewis, and @rafiss)

___
*[`-- commits` line 4 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5y-hnv0nzy3pGKndHn):*
nit: mention the other with custom costs overload too, and end the sentence with a period.
___
*[`pkg/sql/sem/builtins/builtins.go` line 4087 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5xvEg0Af4Zk1bWtOez:-O5xvEg0Af4Zk1bWtOf-:b-1ctt71) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/sql/sem/builtins/builtins.go#L4087)):*
> ```Go
> 			},
> 			Info: "Calculates the Levenshtein distance between two strings. If actual distance is less or equal then max_d, " +
> 				"then returns accurate value of it. Otherwise this function returns value which is greater than max_d. " +
> ```

nit: "then it returns the distance."

nit: "Otherwise this function returns a value greater than max_d."
___
*[`pkg/sql/sem/builtins/builtins.go` line 4088 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5xvPo_7HDZSNAHq1AT:-O5xvPo_7HDZSNAHq1AU:b-q3wsrr) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/sql/sem/builtins/builtins.go#L4088)):*
> ```Go
> 			Info: "Calculates the Levenshtein distance between two strings. If actual distance is less or equal then max_d, " +
> 				"then returns accurate value of it. Otherwise this function returns value which is greater than max_d. " +
> 				"Maximum input length is 255 characters.",
> ```

nit: "The maximum length of the input strings is 255 characters."
___
*[`pkg/sql/sem/builtins/builtins.go` line 4106 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5xxBAW9Dh4N5YW_NKa:-O5xxBAW9Dh4N5YW_NKb:b4zlyt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/sql/sem/builtins/builtins.go#L4106)):*
> ```Go
> 				d := int(tree.MustBeDInt(args[5]))
> 				const maxLen = 255
> 				if len(s) > maxLen || len(t) > maxLen {
> ```

`len(string)` returns the number of bytes. Postgres allows up to 255 characters, which can be more than 255 bytes if some are multi-byte characters. Use `utf8.RuneCountInString(string)` instead. Please also add tests for this in the `fuzzystrmatch` logictest file.
___
*[`pkg/sql/sem/builtins/builtins.go` line 4116 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5xv_3Z0vv5f1cCl4K_:-O5xv_3Z0vv5f1cCl4Ka:b-1d4dp4) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/sql/sem/builtins/builtins.go#L4116)):*
> ```Go
> 				"charge for each edit operation. If actual distance is less or equal then max_d, " +
> 				"then returns accurate value of it. Otherwise this function returns value which is greater than max_d. " +
> 				"Maximum input length is 255 characters.",
> ```

nit: Same edits as above.
___
*[`pkg/util/fuzzystrmatch/leven.go` line 72 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5xvonb9r6SJmOfi5i0:-O5xvonb9r6SJmOfi5i1:b-bgd64b) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/util/fuzzystrmatch/leven.go#L72)):*
> ```Go
> // LevenshteinLessEqualDistance is the standard Levenshtein distance where the cost of
> // insertion, deletion and substitution are all one and we care about distances less than
> // or equal to maxDist.
> ```

nit: Explain the behavior if the distance is longer than `maxD`.
___
*[`pkg/util/fuzzystrmatch/leven.go` line 79 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5xw0-C5mtZkjr2Fu63:-O5xw0-C5mtZkjr2Fu64:b-bgd64b) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/util/fuzzystrmatch/leven.go#L79)):*
> ```Go
> // LevenshteinLessEqualDistanceWithCost is the Levenshtein distance between the source and
> // the target with the given costs of insertion, deletion and substitution and we care about
> // distances less than or equal to maxDist. This approach is adapted from the
> ```

nit: Explain the behavior if the distance is longer than `maxD`.
___
*[`pkg/util/fuzzystrmatch/leven.go` line 88 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5y1qEq5L8xR7obBQyP:-O5y1qEr1zOpaQSODPrc:b-fdkrmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/util/fuzzystrmatch/leven.go#L88)):*
> ```Go
> 	// Converting to a slice of runes allows us to count multi-byte characters
> 	// properly.
> 	s, t := []rune(source), []rune(target)
> ```

Move this allocations to after the block that checks if the bound is impossibly tight, and use`utf8.RuneCountInString` to calculate `lenS` and `lenT` instead.

Also, I think these allocations aren't necessary in all cases. Can you leave this comment:

```
// TODO(mgartner): We should be able to eliminate these allocations in some cases, e.g., when
// the string contains only single-byte characters.
```
___
*[`pkg/util/fuzzystrmatch/leven.go` line 252 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5xytvP34oKyQu_vtIC:-O5xytvP34oKyQu_vtID:bwd22c8) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/util/fuzzystrmatch/leven.go#L252)):*
> ```Go
> }
> 
> func min2(a, b int) int {
> ```

`min` is no a builtin function in Go, so you shouldn't need this.
___
*[`pkg/sql/logictest/testdata/logic_test/fuzzystrmatch` line 17 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5y3RMTEvu0OfJVH0BO:-O5y3RMTEvu0OfJVH0BP:buziobq) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/sql/logictest/testdata/logic_test/fuzzystrmatch#L17)):*
> ```
>   (6, null, null),
>   (7, 'apple', 'apple'),
>   (8, 'a', 'abcd')
> ```

nit: add a case with 2 empty strings, and a case where b is the empty string and a is not.
___
*[`pkg/util/fuzzystrmatch/leven_test.go` line 222 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5y3YdYA4tQB4r-uX9x:-O5y3YdYA4tQB4r-uX9y:b-btvk6u) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/util/fuzzystrmatch/leven_test.go#L222)):*
> ```Go
> 		Expected    int
> 	}{
> 		// Test cases for LevenshteinDistance less than max distance.
> ```

test cases with empty string inputs would be good here too.
___
*[`pkg/util/fuzzystrmatch/leven_test.go` line 401 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5y3sLD-KdpkvV3Fxgu:-O5y3sLD-KdpkvV3Fxgv:bybf79z) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/util/fuzzystrmatch/leven_test.go#L401)):*
> ```Go
> 		{
> 			Source:      "😄 🐃 🐯 🕣 💲 🏜 👞 🔠 🌟 📌",
> 			Target:      "💤 🚈 👨 💩 👲 💽 🔴 🍨 😮 😅",
> ```

Some more thoughtful test cases with multi-byte characters would be good here. These two case don't have any similar characters, so they are entirely additions or deletions.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
i've addressed the remaining comments, and will merge this now. thanks for your contributions! 

bors r+
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-:-OYIv6qA9lNx4LpLpzq2:b2ge9kv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @kev-cao, @mgartner, and @yuzefovich)

___
*[`-- commits` line 4 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5y-hnv0nzy3pGKndHn:-OYItcxBEMwTU95GCv6U:bb7f2mf):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: mention the other with custom costs overload too, and end the sentence with a period.
</blockquote></details>

done
___
*[`pkg/sql/sem/builtins/builtins.go` line 4087 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5xvEg0Af4Zk1bWtOez:-OYIteeeDigVc05XskNz:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/sql/sem/builtins/builtins.go#L4087)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: "then it returns the distance."

nit: "Otherwise this function returns a value greater than max_d."
</blockquote></details>

done
___
*[`pkg/sql/sem/builtins/builtins.go` line 4088 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5xvPo_7HDZSNAHq1AT:-OYItfATCg5vv0eZli8h:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/sql/sem/builtins/builtins.go#L4088)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: "The maximum length of the input strings is 255 characters."
</blockquote></details>

done
___
*[`pkg/sql/sem/builtins/builtins.go` line 4106 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5xxBAW9Dh4N5YW_NKa:-OYIthxr4J_Ox2QxZIcQ:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/sql/sem/builtins/builtins.go#L4106)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

`len(string)` returns the number of bytes. Postgres allows up to 255 characters, which can be more than 255 bytes if some are multi-byte characters. Use `utf8.RuneCountInString(string)` instead. Please also add tests for this in the `fuzzystrmatch` logictest file.
</blockquote></details>

done
___
*[`pkg/sql/sem/builtins/builtins.go` line 4116 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5xv_3Z0vv5f1cCl4K_:-OYItijw-o-W2OhPQESU:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/sql/sem/builtins/builtins.go#L4116)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Same edits as above.
</blockquote></details>

done
___
*[`pkg/util/fuzzystrmatch/leven.go` line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NXVqijP0awPM5bEpyAq:-OYItnb26NZ-Cpu3idxh:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfc4dad72bbc840dc63d8ae597ea96dbdb117661/pkg/util/fuzzystrmatch/leven.go#L192)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Here and in all other comments - can you please end them with a period, `.`?
</blockquote></details>

done
___
*[`pkg/util/fuzzystrmatch/leven.go` line 104 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NYPDQun2bi0ep0ydBsm:-OYItm8e8YMj6nHQobGK:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b87fdea1f382116ea7d29025e4ba2ca46512d4c/pkg/util/fuzzystrmatch/leven.go#L104)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Looks like you forgot to remove this `fmt.Println`
</blockquote></details>

done
___
*[`pkg/util/fuzzystrmatch/leven.go` line 175 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NYPFunL-VYMvo350czU:-OYItmwF7yNiGah-bLZo:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b87fdea1f382116ea7d29025e4ba2ca46512d4c/pkg/util/fuzzystrmatch/leven.go#L175)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I don't see the fast path case, nor any `s_char_len`. 
</blockquote></details>

done
___
*[`pkg/util/fuzzystrmatch/leven.go` line 72 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5xvonb9r6SJmOfi5i0:-OYItjb_7OYGWV4CFP7U:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/util/fuzzystrmatch/leven.go#L72)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Explain the behavior if the distance is longer than `maxD`.
</blockquote></details>

done
___
*[`pkg/util/fuzzystrmatch/leven.go` line 79 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5xw0-C5mtZkjr2Fu63:-OYItkLC5Ittpe390kDi:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/util/fuzzystrmatch/leven.go#L79)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Explain the behavior if the distance is longer than `maxD`.
</blockquote></details>

done
___
*[`pkg/util/fuzzystrmatch/leven.go` line 88 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5y1qEq5L8xR7obBQyP:-OYItl-dCpxCxblfHaPq:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/util/fuzzystrmatch/leven.go#L88)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Move this allocations to after the block that checks if the bound is impossibly tight, and use`utf8.RuneCountInString` to calculate `lenS` and `lenT` instead.

Also, I think these allocations aren't necessary in all cases. Can you leave this comment:

```
// TODO(mgartner): We should be able to eliminate these allocations in some cases, e.g., when
// the string contains only single-byte characters.
```
</blockquote></details>

done
___
*[`pkg/util/fuzzystrmatch/leven.go` line 252 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5xytvP34oKyQu_vtIC:-OYItpFPAJMuuE3hOdc8:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/util/fuzzystrmatch/leven.go#L252)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

`min` is no a builtin function in Go, so you shouldn't need this.
</blockquote></details>

done
___
*[`pkg/sql/logictest/testdata/logic_test/fuzzystrmatch` line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NXVsdXE0oPPeXBmw3sY:-OYItwg177IfSZL2NDSg:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/585f032a689c8f1c329687351ce7e57c4aaf2c65/pkg/sql/logictest/testdata/logic_test/fuzzystrmatch#L47)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Can you add some more test cases here too? Some ideas:

* ASCII strings where the levenshtein distance is less than the given max
* ASCII strings that are identical
</blockquote></details>

done
___
*[`pkg/sql/logictest/testdata/logic_test/fuzzystrmatch` line 40 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NYPIHRR5yY4-AMzjgXh:-OYItuzh4G0VJf-FlmW1:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b87fdea1f382116ea7d29025e4ba2ca46512d4c/pkg/sql/logictest/testdata/logic_test/fuzzystrmatch#L40)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

`levenshtein_less_than` => `levenshtein_less_equal`
</blockquote></details>

done
___
*[`pkg/sql/logictest/testdata/logic_test/fuzzystrmatch` line 49 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-NYPHnL03t1NCFxtmjPV:-OYItyE11_Os6BKS83dF:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b87fdea1f382116ea7d29025e4ba2ca46512d4c/pkg/sql/logictest/testdata/logic_test/fuzzystrmatch#L49)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

You need to insert these values into the table `fuzzystrmatch_table`. That'll require updating all the other tests that use this table. You can automatically do that with the `--rewrite` test flag:

```
./dev test pkg/sql/logictest/tests/local -f TestLogic_fuzzystrmatch --rewrite
```
</blockquote></details>

done
___
*[`pkg/sql/logictest/testdata/logic_test/fuzzystrmatch` line 17 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5y3RMTEvu0OfJVH0BO:-OYItq4u8Ja6-UjczHBZ:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/sql/logictest/testdata/logic_test/fuzzystrmatch#L17)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: add a case with 2 empty strings, and a case where b is the empty string and a is not.
</blockquote></details>

done
___
*[`pkg/util/fuzzystrmatch/leven_test.go` line 222 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5y3YdYA4tQB4r-uX9x:-OYItyv37O0ZXvED-rJ5:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/util/fuzzystrmatch/leven_test.go#L222)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

test cases with empty string inputs would be good here too.
</blockquote></details>

done
___
*[`pkg/util/fuzzystrmatch/leven_test.go` line 401 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/104649#-O5y3sLD-KdpkvV3Fxgu:-OYIu-YaBTPAbIwefKk9:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/71d48634a34f3ff30eafec7de7f2adb897ec01d6/pkg/util/fuzzystrmatch/leven_test.go#L401)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Some more thoughtful test cases with multi-byte characters would be good here. These two case don't have any similar characters, so they are entirely additions or deletions.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->
